### PR TITLE
Add ecdsa self_test

### PR DESCRIPTION
**Pull request by Wenxing-hou**
Reference: #1260

**Both openssl and mbedtsl KAT and PCT test are passed.**

Signed-off-by: Wenxing Hou <wenxing.hou@intel.com>

### CODE REVIEW

File: library/spdm_crypt_lib/CMakeLists.txt [hunk 1]
```diff
@@ -15,6 +16,11 @@ SET(src_spdm_crypt_lib
     fips/libspdm_selftest_hmac.c
     fips/libspdm_selftest_aes_gcm.c
     fips/libspdm_selftest_rsa_ssa.c
+    fips/libspdm_selftest_ecdsa.c
 )
 
+if(CRYPTO STREQUAL "mbedtls")
+    add_definitions(-DLIBSPDM_MBEDTLS_USE=1)
```
Comment (jyao1) [inline library/spdm_crypt_lib/CMakeLists.txt:23]: What if we include 3rd crypto lib?

Comment (Wenxing-hou) [reply library/spdm_crypt_lib/CMakeLists.txt]: Thanks. I have deleted the MACRO. Both mbedtls and openssl have KAT test now.

File: include/hal/library/cryptlib/cryptlib_ec.h [hunk 2]
```diff
@@ -155,6 +155,16 @@ extern bool libspdm_ecdsa_sign(void *ec_context, size_t hash_nid,
...
 
+#if LIBSPDM_FIPS_MODE
+/**
+ * Carries out the EC-DSA signature for fips: hardcode random number.
+ *
+ **/
+extern bool libspdm_ecdsa_sign_fips(void *ec_context, size_t hash_nid,
+                                    const uint8_t *message_hash, size_t hash_size,
+                                    uint8_t *signature, size_t *sig_size);
+#endif/*LIBSPDM_FIPS_MODE*/
```
Comment (jyao1) [inline include/hal/library/cryptlib/cryptlib_ec.h:166]: This does not make sense.

It says: hardcode random number.
But it does not describe how to let the caller input random number value.

Comment (Wenxing-hou) [reply include/hal/library/cryptlib/cryptlib_ec.h:167]: Thanks. I have added a new input parameter(function point) to hardcode the random number.
And both OpenSSL and MbedTLS  tests are passed.

Comment (jyao1) [reply include/hal/library/cryptlib/cryptlib_ec.h:167]: I suggest `libspdm_ecdsa_sign_fips` -> `libspdm_ecdsa_sign_ex`.
`hardcode_random` -> `random_func`.

Comment (Wenxing-hou) [reply include/hal/library/cryptlib/cryptlib_ec.h:179]: Thanks. I have updated the function names and comments.

File: os_stub/cryptlib_mbedtls/pk/ec.c [hunk 3]
```diff
@@ -706,3 +706,121 @@ bool libspdm_ecdsa_verify(void *ec_context, size_t hash_nid,
...
+        break;
+
+    default:
+        return false;
+    }
+
+    mbedtls_mpi_init(&bn_r);
+    mbedtls_mpi_init(&bn_s);
+
+    /*hardcode random number*/
```
Comment (jyao1) [inline os_stub/cryptlib_mbedtls/pk/ec.c:795]: `hardcode random number` -> `retrieve random number`

File: os_stub/cryptlib_openssl/pk/ec.c [hunk 4]
```diff
@@ -846,3 +846,221 @@ bool libspdm_ecdsa_verify(void *ec_context, size_t hash_nid,
 
     return (result == 1);
 }
+
+#if LIBSPDM_FIPS_MODE
+/*setup random number(hardcode)*/
```
Comment (jyao1) [inline os_stub/cryptlib_openssl/pk/ec.c:851]: `setup random number(hardcode)` -> `setup random number`

File: os_stub/cryptlib_openssl/pk/ec.c [hunk 5]
```diff
@@ -846,3 +846,221 @@ bool libspdm_ecdsa_verify(void *ec_context, size_t hash_nid,
...
+{
+    EC_KEY *ec_key;
+    ECDSA_SIG *ecdsa_sig;
+    int32_t openssl_nid;
+    uint8_t half_size;
+    BIGNUM *bn_r;
+    BIGNUM *bn_s;
+    int r_size;
+    int s_size;
+    /* hardcode random number*/
```
Comment (jyao1) [inline os_stub/cryptlib_openssl/pk/ec.c:971]: remove `hardcode random number`

Basic rule: you can hardcode random number in FIPS test, but you cannot hardcode random number in a crypto library.

File: os_stub/cryptlib_openssl/pk/ec.c [hunk 6]
```diff
@@ -846,3 +846,221 @@ bool libspdm_ecdsa_verify(void *ec_context, size_t hash_nid,
...
+        if (hash_size != LIBSPDM_SHA512_DIGEST_SIZE) {
+            return false;
+        }
+        break;
+
+    default:
+        return false;
+    }
+
+    /*hard code random number for ecdsa sign*/
```
Comment (jyao1) [inline os_stub/cryptlib_openssl/pk/ec.c:1033]: `hard code random number for ecdsa sign` -> `retrieve random number for ecdsa sign`

File: os_stub/cryptlib_openssl/pk/ec.c [hunk 7]
```diff
@@ -846,3 +846,221 @@ bool libspdm_ecdsa_verify(void *ec_context, size_t hash_nid,
...
+
+    /* Preallocate space */
+    order_bits = BN_num_bits(order);
+    if (!BN_set_bit(k, order_bits)
+        || !BN_set_bit(r, order_bits)
+        || !BN_set_bit(X, order_bits)) {
+        goto err;
+    }
+
+    /*hard code random number*/
```
Comment (jyao1) [inline os_stub/cryptlib_openssl/pk/ec.c:903]: remove the comment

File: os_stub/cryptlib_openssl/pk/ec.c (hunk 5 above)
Comment (Wenxing-hou) [reply os_stub/cryptlib_openssl/pk/ec.c]: Thanks. I will follow the basic rule.

File: os_stub/cryptlib_openssl/pk/ec.c [hunk 8]
```diff
@@ -846,3 +846,221 @@ bool libspdm_ecdsa_verify(void *ec_context, size_t hash_nid,
...
+    EC_KEY *ec_key;
+    ECDSA_SIG *ecdsa_sig;
+    int32_t openssl_nid;
+    uint8_t half_size;
+    BIGNUM *bn_r;
+    BIGNUM *bn_s;
+    int r_size;
+    int s_size;
+    /*random number*/
+    uint8_t random[32];
```
Comment (jyao1) [inline os_stub/cryptlib_openssl/pk/ec.c:972]: this random number must be cleared before return.

Comment (Wenxing-hou) [reply os_stub/cryptlib_openssl/pk/ec.c:972]: Yes. In ECDSA, the random number is the as import as the secret.
Thanks. I have fix it.

File: os_stub/cryptlib_openssl/pk/ec.c [hunk 9]
```diff
@@ -846,3 +846,230 @@ bool libspdm_ecdsa_verify(void *ec_context, size_t hash_nid,
...
+    bool result;
+
+    result = true;
+
+    BIGNUM *kinv = NULL;
+    BIGNUM *rp = NULL;
```
Comment (jyao1) [inline os_stub/cryptlib_openssl/pk/ec.c:978]: you need BN_clear_free kinv and rp.

**Pull request merged**